content/blog/2021-12-07-ml-experiment-versioning.md

```diff
@@ -0,0 +1,293 @@
+steps or stages run. Each stage should include info about the commands,
+dependencies, and outputs needed to reproduce each stage and to piece together
+the pipeline. Like other artifacts, metadata about the pipeline is saved in
+plain-text files to track in the code repo.
+
+```dvc
+$ cat dvc.yaml
+stages:
+  default:
+    cmd: python src/train.py
+    deps:
+    - data/images
+    - src/train.py
+    params:
+    - model
+    - train
+    outs:
+    - models
+    metrics:
+    - metrics.json:
+        cache: false
+    plots:
+    - logs.csv:
+        cache: false
+```
+
+Codifying the entire experiment pipeline is a good first step towards
+reproducibility, but it still leaves the user to execute that pipeline, and ML
+pipelines often take too long to reproduce on the fly. Versioning the pipeline
+means tracking changes to it. If a pipeline stage's commands and dependencies
+have not changed, its outputs may be recovered so that it does not have to be
+executed again.
+
+```dvc
+$ dvc exp run
+'data/images.dvc' didn't change, skipping
+Stage 'default' didn't change, skipping
+
+Reproduced experiment(s): exp-333c9
+Experiment results have been applied to your workspace.
+
+To promote an experiment to a Git branch run:
+
+        dvc exp branch <exp> <branch>
+```
+
+# Distributed Experiments
+
+Experiment tracking tools log experiments to a central database and show them in
+a dashboard. This enables comparison that would be difficult in Git. Distributed
+experiments are recorded in the local code repo like leaves or twigs growing
+from Git branches.
+
+```dvc
+$ dvc exp show
+┏━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┓
+┃ Experiment              ┃ Created      ┃    loss ┃    acc ┃ train.epochs ┃ model.conv_units ┃
+┡━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━┩
+│ workspace               │ -            │ 0.25183 │ 0.9137 │ 10           │ 64
+│
+│ mybranch                │ Oct 23, 2021 │       - │      - │ 10           │ 16
+│
+│ ├── 9a4ff1c [exp-333c9] │ 10:40 AM     │ 0.25183 │ 0.9137 │ 10           │ 64
+│
+│ ├── 138e6ea [exp-55e90] │ 10:28 AM     │ 0.25784 │ 0.9084 │ 10           │ 32
+│
+│ └── 51b0324 [exp-2b728] │ 10:17 AM     │ 0.25829 │ 0.9058 │ 10           │ 16
+│
+└─────────────────────────┴──────────────┴─────────┴────────┴──────────────┴──────────────────┘
+```
+
+You may push and pull what you want to share. When collaborating, only a small
+selection of experiments may be worth sharing, and you may not know until after
+you run and compare them. Pushing an experiment shares all the experiment
+information and code so that they stay in sync and are shared with the same
+people, and it's already organized using Git branches.
+
+```dvc
+$ dvc exp push origin exp-333c9
+Pushed experiment 'exp-333c9'to Git remote 'origin'.
+```
+
+While Git branching is a flexible way to organize experiments, there are often
```

Comment:
It might worth to write a separate section about ephemeral branches with the link to Peter's blog post about custom references. This is an interesting technical info can attract attention of some folks.

# Experiments as ephemeral Git branches

To keep track of all your experiments you need to support references to all of them. Git branches is an easy way of doing this. However, using Git branches for hundreds of experiments can easily over pollute namespace of your branches - you don't want to see 100s of branches as an output of `git branch` command.

We implemented a special machinery in DVC that we called "ephemeral branches" or experiment references in
order to separate Git branches and experiment branches. This machinery is implemented with [Git 
custom references](https://git-scm.com/book/en/v2/Git-Internals-Git-References).
DVC just introduces a new type of custom references `exps` in addition to the standard ones - `branch` (or `heads`), `remotes`, `tags`.

Git custom references is an advance Git functionality that is used by some services. For example, GitHub stores
pull requests as custom references (proof or code example is needed). DVC uses similar approach to help data scientists and ML engineers manage experiments.

Reply:
The proof with code example: https://stackoverflow.com/questions/70141055/git-fetch-pull-requests-from-github

tl;dr; `git ls-remote origin | grep "refs/pull"`

Reply:
Hmm, this is the only comment where I'm not convinced 😄 . This could be interesting technical info for some folks, but does it belong in this post? Technical info isn't the focus of this post, and we can link to Peter's post for those who are interested. It doesn't seem to directly address how experiment versioning improves on experiment tracking (I would argue it's less relevant than concepts like data versioning and pipelines that we are removing). It seems more about how we improve the typical git workflow, which we have generally deprioritized to focus on comparisons to experiment tracking tools. It also doesn't really tie in to other concepts or examples in the post, which is already starting to tilt back towards more abstract and conceptual.

Reply:
It's up to you.

I agree that the current version does not fit to the current post. I probably presented it in a wrong way. It should be (if you decide to go with it) more lightweight. Title should be like "Under the hood" or "Git internals". It might be just a simple, easy-to-understand paragraph when we should mention that we use some Git internals, like GitHub does.